compiler/rustc_hir_analysis/src/check/mod.rs

```diff
@@ -319,41 +319,56 @@ fn bounds_from_generic_predicates<'tcx>(
             _ => {}
         }
     }
-    let generics = if types.is_empty() {
-        "".to_string()
-    } else {
-        format!(
-            "<{}>",
-            types
-                .keys()
-                .filter_map(|t| match t.kind() {
-                    ty::Param(_) => Some(t.to_string()),
-                    // Avoid suggesting the following:
-                    // fn foo<T, <T as Trait>::Bar>(_: T) where T: Trait, <T as Trait>::Bar: Other {}
-                    _ => None,
-                })
-                .collect::<Vec<_>>()
-                .join(", ")
-        )
-    };
+
     let mut where_clauses = vec![];
+    let mut types_str = vec![];
     for (ty, bounds) in types {
-        where_clauses
-            .extend(bounds.into_iter().map(|bound| format!("{}: {}", ty, tcx.def_path_str(bound))));
-    }
-    for projection in &projections {
-        let p = projection.skip_binder();
-        // FIXME: this is not currently supported syntax, we should be looking at the `types` and
-        // insert the associated types where they correspond, but for now let's be "lazy" and
-        // propose this instead of the following valid resugaring:
-        // `T: Trait, Trait::Assoc = K` → `T: Trait<Assoc = K>`
-        where_clauses.push(format!("{} = {}", tcx.def_path_str(p.projection_ty.def_id), p.term));
+        if let ty::Param(_) = ty.kind() {
+            let mut bounds_str = vec![];
+            for bound in bounds {
+                let mut projections_str = vec![];
+                for projection in &projections {
+                    let p = projection.skip_binder();
+                    let alias_ty = p.projection_ty;
+                    if bound == tcx.parent(alias_ty.def_id) && alias_ty.self_ty() == ty {
+                        let def_path = tcx.def_path(alias_ty.def_id);
+                        projections_str.push(format!(
+                            "{} = {}",
+                            def_path.data.last().unwrap(),
```

Comment:
I'm uncertain whether it's safe to use ﻿`unwrap` in this context

Reply:
I think it's easier to just get `tcx.item_name(alias_ty.def_id)` to get the name of the associated type

Reply:
Should RPITIT be skipped here?

Reply:
RPITITs shouldn't show up here anyways, or else I'd consider that a bug in the caller?

Reply:
> RPITITs shouldn't show up here anyways

Sorry, I am slightly perplexed by this case. To clarify, considering the following code:

```rs
trait TraitA {
     type A;
}

trait TraitB<T> {
    fn f(&self) -> impl TraitA<Item = T>;
}

struct X;
impl TraitB<u8> for X {}
```

Are you suggesting that we should avoid displaying the help message for `﻿implement the missing item: 'fn f(&self) -> impl TraitB<Item = u8> { todo!() }' `?